File: 03-stream-processing/solution-proposal/scala/src/main/scala/com/becomedataengineer/exercise3/DeadLetterQueueJob.scala
package com.becomedataengineer.exercise3

import org.apache.spark.sql.{SparkSession, functions}

object DeadLetterQueueJob {

  def main(args: Array[String]): Unit = {
    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()
    import sparkSession.implicits._

    val inputStream = sparkSession.readStream.format("kafka")
      .options(Map(
        "kafka.bootstrap.servers" -> "localhost:29092",
        "subscribe" -> "visits_raw_for_dlq",
        "startingOffsets" -> "EARLIEST"
      )).load()

    val validAndInvalidVisits = inputStream
      .select(functions.from_json($"value".cast("string"), UserVisit.Schema).as("value"))
      .selectExpr("value.*", "key")
      .as[UserVisit]
      .map(visit => {
        // Whoops, the visitedPage is an `Option[String]`, so it can be missing!
        // We call it with a direct `.get`, making the code fail
        // Let's fix it by changing the function and catch any error
        try {
          VisitPageMappingResult(s"""{"page":  "${visit.visitedPage.get}"}""", visit.key, "valid_data")
        } catch {
          case e: Exception => {
            VisitPageMappingResult(visit.asJson, visit.key, "dead_letter")
          }
        }
      })

    val writeQuery = validAndInvalidVisits.select("value", "key", "topic")
      .writeStream.format("kafka")
      .options(
        Map(
          "kafka.bootstrap.servers" -> "localhost:29092",
          "checkpointLocation" -> CheckpointDir
        )
      ).start()

    writeQuery.awaitTermination()
  }

}

case class VisitPageMappingResult(value: String, key: String, topic: String)

File: 03-stream-processing/solution-proposal/python/exercise3/dead_letter_queue_job.py
import pyspark.sql.functions as F
from pyspark.sql import SparkSession

from exercise3.mappers import get_visited_page

spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.4.0') \
    .getOrCreate()

input_data_stream = spark.readStream \
    .option("kafka.bootstrap.servers", "localhost:29092") \
    .option("subscribe", "visits_raw_for_dlq") \
    .option("startingOffsets", "EARLIEST") \
    .format("kafka") \
    .load()


valid_and_invalid_visits = input_data_stream.withColumn("payload", get_visited_page(F.col("value")))


write_query = valid_and_invalid_visits.select("payload.value", "payload.topic", "key")\
    .writeStream.format("kafka")\
    .option("checkpointLocation", "/tmp/bde/module5/homework/e3/checkpoints")\
    .option("kafka.bootstrap.servers", "localhost:29092")\
    .start()

write_query.awaitTermination()
